extensions/cornerstone/src/services/ViewportService/CornerstoneViewportService.ts

```diff
@@ -177,29 +242,27 @@ class CornerstoneViewportService implements IViewportService {
     viewportIndex: number,
     viewportData: StackViewportData | VolumeViewportData,
     publicViewportOptions: PublicViewportOptions,
-    publicDisplaySetOptions: DisplaySetOptions[]
+    publicDisplaySetOptions: DisplaySetOptions[],
+    presentation?: Presentation
   ): void {
     const renderingEngine = this.getRenderingEngine();
-    const viewportInfo = this.viewportsInfo.get(viewportIndex);
-
-    if (!publicViewportOptions.viewportId) {
-      publicViewportOptions.viewportId = this.getViewportId(viewportIndex);
+    const viewportId =
+      publicViewportOptions.viewportId || this.getViewportId(viewportIndex);
+    if (!viewportId) {
+      throw new Error('Must define viewportId externally');
     }
 
-    let viewportId = viewportInfo.getViewportId();
+    const viewportInfo = this.viewportsById.get(viewportId);
 
-    // if currently there is a viewport with the viewportId, but it is not the same
-    // as the one we are trying to set, we need to disable the old one
-    // and enable the new one, we could ideally change the name of the viewportId
-    // but the viewportId is an integral part in renderers map, tools svg cache
-    // etc. which would require a lot of refactoring, for now we will just disable
-    // the old one and enable the new one at the end of this function
-    let newViewportId = null;
-    if (publicViewportOptions?.viewportId !== viewportId) {
-      newViewportId = publicViewportOptions.viewportId;
-      viewportInfo.setViewportId(newViewportId);
+    if (!viewportInfo) {
+      throw new Error('Viewport info not defined');
+    }
 
-      renderingEngine.disableElement(viewportId);
+    // If the viewport has moved index, then record the new index
+    if (viewportInfo.viewportIndex !== viewportIndex) {
+      this.viewportsInfo.delete(viewportInfo.viewportIndex);
+      this.viewportsInfo.set(viewportIndex, viewportInfo);
+      viewportInfo.viewportIndex = viewportIndex;
     }
```

Comment:
MPR is not working because of the same exact comment I put in the old implementation, here it is. Basically we can't just re name the viewports as far as I know, since a stack turning to MPR should disable stack. Am I missing something?


```
   // if currently there is a viewport with the viewportId, but it is not the same
    // as the one we are trying to set, we need to disable the old one
    // and enable the new one, we could ideally change the name of the viewportId
    // but the viewportId is an integral part in renderers map, tools svg cache
    // etc. which would require a lot of refactoring, for now we will just disable
    // the old one and enable the new one at the end of this function
  ```

Reply:
No, the disabling is handled by the removing only - this is just a move for an existing viewport.